Implement TCP/IP

By this spec:
http://wiretap.area.com/Gopher/pub/tsoft/trumpet-abi/tcp201.specification

Anyone? :)

Unix part doesn't seem to build:
```
$ LC_ALL=C make
makefile:88: *** missing separator.  Stop.

$ LC_ALL=C make -f Makefile 
cc -DNEED_SELECT_H    -c tshost.c
tshost.c:45:14: error: conflicting types for 'sys_errlist'
 extern char *sys_errlist[];
              ^~~~~~~~~~~
In file included from /usr/include/stdio.h:787,
                 from tshost.c:17:
/usr/include/bits/sys_errlist.h:27:26: note: previous declaration of 'sys_errlist' was here
 extern const char *const sys_errlist[];
                          ^~~~~~~~~~~
<and many more errors follow>
```

Also we need to think up the use case.
IMHO if we have trumpet API implemented in
dosemu2, there is no use-case for this, but I'd
be very glad to be wrong.

--- a/src/base/misc/utilities.c
+++ b/src/base/misc/utilities.c
@@ -1334,11 +1334,10 @@ char *concat_strings(char *dst, const char *pref, const char *suff)
   return ret;
 }
 
-int handle_timeout(uint16_t to,
+static int do_to(uint32_t to, uint32_t inf, hitimer_t end,
     enum CbkRet (*cbk)(int, void *, int, int *),
     int arg, void *arg2, int arg3, int *r_err)
 {
-    hitimer_t end = GETusTIME(0) + TICKtoUS(to * 65536);
     int first = 1;
     enum CbkRet cbr;
 
@@ -1351,10 +1350,26 @@ int handle_timeout(uint16_t to,
         if (cbr == CBK_DONE)
             break;
         first = 0;
-    } while (to && (to == 0xffff || GETusTIME(0) < end));
+    } while (to && (to == inf || GETusTIME(0) < end));
     if (cbr != CBK_DONE)
         return 1;
     return 0;
+}
+
+int handle_timeout(uint16_t to,
+    enum CbkRet (*cbk)(int, void *, int, int *),
+    int arg, void *arg2, int arg3, int *r_err)
+{
+    hitimer_t end = GETusTIME(0) + TICKtoUS(to * 65536);
+    return do_to(to, 0xffff, end, cbk, arg, arg2, arg3, r_err);
+}
+
+int handle_timeout_us(uint32_t to,
+    enum CbkRet (*cbk)(int, void *, int, int *),
+    int arg, void *arg2, int arg3, int *r_err)
+{
+    hitimer_t end = GETusTIME(0) + to;
+    return do_to(to, 0xffffffff, end, cbk, arg, arg2, arg3, r_err);
 }
 
 int send_fd(int usock, int fd_tx)

--- a/src/dosext/dpmi/Makefile
+++ b/src/dosext/dpmi/Makefile
@@ -1,7 +1,8 @@
 top_builddir=../../..
 include $(top_builddir)/Makefile.conf
 
-CFILES = dpmi.c memory.c emu-ldt.c msdoshlp.c vxd.c coopth_pm.c dpmi_api.c
+CFILES = dpmi.c memory.c emu-ldt.c msdoshlp.c vxd.c coopth_pm.c dpmi_api.c \
+  vxd_sock.c
 SFILES = dpmisel.S
 ALL_CPPFLAGS += -DDOSEMU
 

--- a/src/dosext/dpmi/dpmisel.S
+++ b/src/dosext/dpmi/dpmisel.S
@@ -240,6 +240,10 @@ DPMI_VXD_APM:
 DPMI_VXD_VTDAPI:
 	hlt
 	lretl
+	.globl	DPMI_VXD_SOCK
+DPMI_VXD_SOCK:
+	hlt
+	lretl
 	.globl	DPMI_VXD_end
 DPMI_VXD_end:
 

--- a/src/dosext/dpmi/dpmisel_offsets.h
+++ b/src/dosext/dpmi/dpmisel_offsets.h
@@ -44,6 +44,7 @@ extern const unsigned    DPMI_VXD_CONFIGMG;
 extern const unsigned    DPMI_VXD_ENABLE;
 extern const unsigned    DPMI_VXD_APM;
 extern const unsigned    DPMI_VXD_VTDAPI;
+extern const unsigned    DPMI_VXD_SOCK;
 extern const unsigned    DPMI_VXD_end;
 extern const unsigned    MSDOS_pmc_start;
 extern const unsigned    MSDOS_fault;

--- a/src/dosext/dpmi/vxd.c
+++ b/src/dosext/dpmi/vxd.c
@@ -106,6 +106,11 @@ void get_VXD_entry(cpuctx_t *scp )
 	    _es = dpmi_sel();
 	    _edi = DPMI_SEL_OFF(DPMI_VXD_VTDAPI);
 	    break;
+	case 0x1235:
+	    D_printf("DPMI: SOCK VxD entry point requested\n");
+	    _es = dpmi_sel();
+	    _edi = DPMI_SEL_OFF(DPMI_VXD_SOCK);
+	    break;
 	default:
 	    D_printf("DPMI: ERROR: Unsupported VxD 0x%x\n", _LWORD(ebx));
 	    /* no entry point */
@@ -1872,5 +1877,9 @@ void vxd_call(cpuctx_t *scp)
     } else if (_eip==1+DPMI_SEL_OFF(DPMI_VXD_VTDAPI)) {
       D_printf("DPMI: VTDAPI VxD called, ax=%#x\n", _LWORD(eax));
       VXD_TimerAPI(scp);
+
+    } else if (_eip==1+DPMI_SEL_OFF(DPMI_VXD_SOCK)) {
+      D_printf("DPMI: SOCK VxD called, ax=%#x\n", _LWORD(eax));
+      VXD_Sock(scp);
     }
 }

--- a/src/dosext/dpmi/vxd.h
+++ b/src/dosext/dpmi/vxd.h
@@ -6,3 +6,5 @@
 
 void get_VXD_entry(cpuctx_t *scp );
 void vxd_call(cpuctx_t *scp);
+
+void VXD_Sock(cpuctx_t *scp);

--- a/src/dosext/dpmi/vxd_sock.c
+++ b/src/dosext/dpmi/vxd_sock.c
@@ -1,0 +1,403 @@
+/*
+ *  Copyright (C) 2025  @stsp
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <stdio.h>
+#include <errno.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <fcntl.h>
+#include "cpu.h"
+#include "dosemu_debug.h"
+#include "utilities.h"
+#include "emudpmi.h"
+#include "emu.h"
+#include "vxd.h"
+
+enum { SOCK_VER, SOCK_OPEN, SOCK_CLOSE, SOCK_BIND, SOCK_SENDTO,
+       SOCK_RECVFROM, SOCK_SELECT, SOCK_CONNECT, SOCK_SEND, SOCK_RECV,
+       SOCK_LISTEN, SOCK_ACONNECT, SOCK_ACCEPT, SOCK_GETSOCKNAME,
+       SOCK_GETPEERNAME, SOCK_NBM, SOCK_GETFDS };
+
+struct sock_s {
+    int fd;
+    unsigned int used:1;
+};
+#define SOCK_MAX 32
+static struct sock_s socks[SOCK_MAX];
+static int num_socks;
+
+enum { P_TCP = 6, P_UDP = 17 };
+
+static struct sock_s *sock_alloc(void)
+{
+    struct sock_s *ret;
+    int i;
+
+    for (i = 0; i < num_socks; i++) {
+        if (!socks[i].used)
+            break;
+    }
+    if (i == num_socks) {
+        if (num_socks >= SOCK_MAX) {
+            error("socket table overflow\n");
+            return NULL;
+        }
+        ret = &socks[num_socks++];
+        assert(!ret->used);
+        ret->used = 1;
+        return ret;
+    }
+    ret = &socks[i];
+    ret->used = 1;
+    return ret;
+}
+
+static void sock_free(struct sock_s *sock)
+{
+    sock->used = 0;
+    while (num_socks && !socks[num_socks - 1].used)
+        num_socks--;
+}
+
+struct sel_arg {
+    fd_set rfds;
+    fd_set wfds;
+    fd_set efds;
+};
+
+static enum CbkRet select_cb(int arg0, void *arg, int nfds, int *r_err)
+{
+    struct sel_arg *sel = arg;
+    struct timeval to = {};
+    int rc = select(nfds, &sel->rfds, &sel->wfds, &sel->efds, &to);
+    *r_err = rc;
+    if (rc == -1 && errno != EAGAIN) {
+        error("select(): %s\n", strerror(errno));
+        return CBK_ERR;
+    }
+    if (rc > 0)
+        return CBK_DONE;
+    return CBK_CONT;
+}
+
+#define SOCK_IDX(s) (s - socks)
+
+#define CSOCK_ERR_INVALID_PARAM       0x00000100
+#define CSOCK_ERR_NO_MEMORY           0x00000200
+#define CSOCK_ERR_INVALID_SOCKET      0x00000300
+#define CSOCK_ERR_ALREADY_BOUND       0x00000400
+#define CSOCK_ERR_NOT_BOUND           0x00000500
+#define CSOCK_ERR_ACCESS              0x00000600
+#define CSOCK_ERR_INTERNAL            0x00000700
+#define CSOCK_ERR_FD_INUSE            0x00000800
+#define CSOCK_ERR_INFINITE_WAIT       0x00000900
+#define CSOCK_ERR_NOT_CONNECTED       0x00000a00
+#define CSOCK_ERR_WOULD_BLOCK         0x00000b00
+#define CSOCK_ERR_NOT_LISTENING       0x00000c00
+
+void VXD_Sock(cpuctx_t *scp)
+{
+    struct sock_s *sock;
+    int rc;
+
+    _eflags &= ~CF;
+    switch(_eax) {
+        case SOCK_VER:
+            if (_edi && _edi != 0xffff) {
+                rc = tcp_get_driver_info(SEL_ADR(_ds, _esi), _edi);
+                if (rc) {
+                    _eflags |= CF;
+                    break;
+                }
+            }
+            _eax = 2;
+            break;
+
+        case SOCK_OPEN:
+            sock = sock_alloc();
+            if (!sock) {
+                _eax = CSOCK_ERR_FD_INUSE;
+                _eflags |= CF;
+                break;
+            }
+            if (_ebx == P_TCP)
+                sock->fd = socket(AF_INET, SOCK_STREAM, 0);
+            else
+                sock->fd = socket(AF_INET, SOCK_DGRAM, 0);
+            if (sock->fd == -1) {
+                perror("socket()");
+                sock_free(sock);
+                _eax = CSOCK_ERR_INTERNAL;
+                _eflags |= CF;
+                break;
+            }
+            _edi = SOCK_IDX(sock);
+            _eax = 0;
+            break;
+
+#define TCP_PROLOG0 \
+            if (_edi >= num_socks || !socks[_edi].used) { \
+                error("socket %i not allocated\n", _edi); \
+                _eax = CSOCK_ERR_INVALID_SOCKET; \
+                _eflags |= CF; \
+                break; \
+            } \
+            sock = &socks[_edi]
+
+        case SOCK_CLOSE:
+            TCP_PROLOG0;
+            close(sock->fd);
+            sock_free(sock);
+            _eax = 0;
+            break;
+
+#define TCP_PROLOG \
+            struct sockaddr_in sa; \
+            TCP_PROLOG0; \
+            sa.sin_addr.s_addr = _ebx; \
+            sa.sin_port = _edx; \
+            sa.sin_family = AF_INET
+
+#define TCP_IO(n, op, c) \
+            rc = op; \
+            switch (rc) { \
+            case 0: \
+                error(n": EOF\n"); \
+                _eax = CSOCK_ERR_NOT_CONNECTED; \
+                _eflags |= CF; \
+                break; \
+            case -1: \
+                error(n": %s\n", strerror(errno)); \
+                _eax = CSOCK_ERR_WOULD_BLOCK; \
+                _eflags |= CF; \
+                break; \
+            default: \
+                _eax = 0; \
+                _ecx = rc; \
+                c \
+                break; \
+            }
+
+        case SOCK_BIND: {
+            TCP_PROLOG;
+
+            rc = bind(sock->fd, (struct sockaddr *)&sa, sizeof(sa));
+            if (rc) {
+                error("TCP bind to port %i: %s\n", _edx, strerror(errno));
+                _eax = CSOCK_ERR_ALREADY_BOUND;
+                _eflags |= CF;
+                break;
+            }
+            _eax = 0;
+            break;
+        }
+
+        case SOCK_SENDTO: {
+            TCP_PROLOG;
+            TCP_IO("UDP sendto", sendto(sock->fd, SEL_ADR(_ds, _esi), _ecx, 0,
+                    (struct sockaddr *)&sa, sizeof(sa)),);
+            break;
+        }
+
+        case SOCK_RECVFROM: {
+            struct sockaddr_in sa;
+            socklen_t len = sizeof(sa);
+            TCP_PROLOG0;
+            TCP_IO("UDP recvfrom", recvfrom(sock->fd, SEL_ADR(_ds, _esi), _ecx, 0,
+                    (struct sockaddr *)&sa, &len),
+                _ebx = sa.sin_addr.s_addr;
+                _edx = sa.sin_port;
+            );
+            break;
+        }
+
+        case SOCK_SELECT: {
+            struct sel_arg sel;
+            int i, err = 0, nfds = -1;
+
+            for (i = 0; i < num_socks; i++) {
+                sock = &socks[i];
+                int used = 0;
+                if (!sock->used)
+                    continue;
+                if ((1 << i) & _ebx) {
+                    FD_SET(sock->fd, &sel.rfds);
+                    used++;
+                }
+                if ((1 << i) & _ecx) {
+                    FD_SET(sock->fd, &sel.wfds);
+                    used++;
+                }
+                if ((1 << i) & _edx) {
+                    FD_SET(sock->fd, &sel.efds);
+                    used++;
+                }
+                if (used && nfds < sock->fd)
+                    nfds = sock->fd;
+            }
+            nfds++;
+            switch (handle_timeout_us(_edi, select_cb,
+                    -1, &sel, nfds, &err)) {
+                case 0:
+                    _ebx = _ecx = _edx = 0;
+                    _eax = err;
+                    for (i = 0; i < num_socks; i++) {
+                        sock = &socks[i];
+                        if (FD_ISSET(sock->fd, &sel.rfds))
+                            _ebx |= 1 << i;
+                        if (FD_ISSET(sock->fd, &sel.wfds))
+                            _ecx |= 1 << i;
+                        if (FD_ISSET(sock->fd, &sel.efds))
+                            _edx |= 1 << i;
+                    }
+                    break;
+                case 1:
+                    _eax = CSOCK_ERR_INTERNAL;
+                    _eflags |= CF;
+                    break;
+            }
+            break;
+        }
+
+        case SOCK_CONNECT: {
+            TCP_PROLOG;
+
+            rc = connect(sock->fd, (struct sockaddr *)&sa, sizeof(sa));
+            if (rc) {
+                error("TCP connect: %s\n", strerror(errno));
+                _eax = CSOCK_ERR_WOULD_BLOCK;
+                _eflags |= CF;
+                break;
+            }
+            _eax = 0;
+            break;
+        }
+
+        case SOCK_SEND:
+            TCP_PROLOG0;
+            TCP_IO("TCP send", send(sock->fd, SEL_ADR(_ds, _esi), _ecx, 0),);
+            break;
+
+        case SOCK_RECV:
+            TCP_PROLOG0;
+            TCP_IO("TCP recv", recv(sock->fd, SEL_ADR(_ds, _esi), _ecx, 0),);
+            break;
+
+        case SOCK_LISTEN:
+            TCP_PROLOG0;
+
+            rc = listen(sock->fd, _ecx);
+            if (rc) {
+                error("TCP listen: %s\n", strerror(errno));
+                _eax = CSOCK_ERR_INTERNAL;
+                _eflags |= CF;
+                break;
+            }
+            _eax = 0;
+            break;
+
+        case SOCK_ACONNECT:
+            // NIY
+            break;
+
+        case SOCK_ACCEPT: {
+            TCP_PROLOG0;
+            struct sock_s *s;
+            struct sockaddr_in sa;
+            socklen_t len = sizeof(sa);
+
+            rc = accept(sock->fd, (struct sockaddr *)&sa, &len);
+            if (rc <= 0) {
+                error("TCP accept: %s\n", strerror(errno));
+                _eax = CSOCK_ERR_WOULD_BLOCK;
+                _eflags |= CF;
+                break;
+            }
+            s = sock_alloc();
+            if (!s) {
+                close(rc);
+                _eax = CSOCK_ERR_INTERNAL;
+                _eflags |= CF;
+                break;
+            }
+            s->fd = rc;
+            _edx = SOCK_IDX(s);
+            _eax = 0;
+            _ebx = sa.sin_port;
+            _ecx = sa.sin_addr.s_addr;
+            break;
+        }
+
+        case SOCK_GETSOCKNAME: {
+            TCP_PROLOG0;
+            struct sockaddr_in sa;
+            socklen_t len = sizeof(sa);
+
+            rc = getsockname(sock->fd, (struct sockaddr *)&sa, &len);
+            if (rc) {
+                error("TCP getsockname: %s\n", strerror(errno));
+                _eax = CSOCK_ERR_INTERNAL;
+                _eflags |= CF;
+                break;
+            }
+            _eax = 0;
+            _ebx = sa.sin_port;
+            _ecx = sa.sin_addr.s_addr;
+            break;
+        }
+
+        case SOCK_GETPEERNAME: {
+            TCP_PROLOG0;
+            struct sockaddr_in sa;
+            socklen_t len = sizeof(sa);
+
+            rc = getpeername(sock->fd, (struct sockaddr *)&sa, &len);
+            if (rc) {
+                error("TCP getpeername: %s\n", strerror(errno));
+                _eax = CSOCK_ERR_INTERNAL;
+                _eflags |= CF;
+                break;
+            }
+            _eax = 0;
+            _ebx = sa.sin_port;
+            _ecx = sa.sin_addr.s_addr;
+            break;
+        }
+
+        case SOCK_NBM:
+            TCP_PROLOG0;
+
+            if (!_ecx)
+                _edx = !!(fcntl(sock->fd, F_GETFL) & O_NONBLOCK);
+            else
+                fcntl(sock->fd, F_SETFL, _edx ? O_NONBLOCK : 0);
+            _eax = 0;
+            break;
+
+        case SOCK_GETFDS: {
+            int i;
+
+            _eax = 0;
+            for (i = 0; i < num_socks; i++) {
+                sock = &socks[i];
+                if (!sock->used)
+                    continue;
+                _eax |= 1 << i;
+            }
+            break;
+        }
+    }
+}

--- a/src/dosext/net/tcp.c
+++ b/src/dosext/net/tcp.c
@@ -354,6 +354,20 @@ static int get_driver_info(struct driver_info_rec *di_out)
     return ret;
 }
 
+int tcp_get_driver_info(void *di_out, int len)
+{
+    struct driver_info_rec di;
+    int rc;
+
+    if (len > sizeof(di))
+        return -1;
+    rc = get_driver_info(&di);
+    if (rc)
+        return rc;
+    memcpy(di_out, &di, len);
+    return 0;
+}
+
 static enum CbkRet conn_cb(int fd, void *sa, int len, int *r_err)
 {
     int err = connect(fd, sa, len);

--- a/src/include/emu.h
+++ b/src/include/emu.h
@@ -509,6 +509,7 @@ extern void HMA_init(void);
 extern void hardware_run(void);
 extern int register_exit_handler(void (*handler)(void));
 void tcp_helper(struct vm86_regs *);
+int tcp_get_driver_info(void *di_out, int len);
 void ipx_helper(struct vm86_regs *);
 void free_fonts(void);
 int emulate(int argc, char **argv, char * const *envp);

--- a/src/include/utilities.h
+++ b/src/include/utilities.h
@@ -49,6 +49,9 @@ enum CbkRet { CBK_DONE, CBK_CONT, CBK_ERR };
 int handle_timeout(uint16_t to,
     enum CbkRet (*cbk)(int, void *, int, int *),
     int arg, void *arg2, int arg3, int *r_err);
+int handle_timeout_us(uint32_t to,
+    enum CbkRet (*cbk)(int, void *, int, int *),
+    int arg, void *arg2, int arg3, int *r_err);
 
 int send_fd(int usock, int fd_tx);
 int recv_fd(int sock);

